jujuclient/file.go

```diff
@@ -547,6 +544,57 @@ func updateModels(
 	return nil
 }
 
+// SetModels implements ModelUpdater.
+func (s *store) SetModels(controllerName string, models map[string]ModelDetails) error {
+	if err := ValidateControllerName(controllerName); err != nil {
+		return errors.Trace(err)
+	}
+
+	modelNames := set.NewStrings()
+	for modelName, details := range models {
+		if err := ValidateModel(modelName, details); err != nil {
+			return errors.Trace(err)
+		}
+		modelNames.Add(modelName)
+	}
+
+	releaser, err := s.acquireLock()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer releaser.Release()
+
+	err = updateModels(controllerName, func(storedModels *ControllerModels) (bool, error) {
+		changed := len(storedModels.Models) != len(models)
+		// Add or update controller models based on a new collection.
+		for modelName, details := range models {
+			oldDetails, ok := storedModels.Models[modelName]
+			if ok && details == oldDetails {
+				continue
+			}
+			storedModels.Models[modelName] = details
+			changed = true
+		}
+		// Delete models that are not in the new collection.
+		for modelName, _ := range storedModels.Models {
+			if _, ok := models[modelName]; !ok {
+				delete(storedModels.Models, modelName)
+				changed = true
```

Comment:
changed = true is redundant here. if we got here, then either:

 - len(storedModels.Models) != len(models), in which case changed == true
 - len(storedModels.Models) == len(models), which implies that the same number of models were added as were removed